src/api_client/tests/test_client_api.py

```diff
@@ -0,0 +1,202 @@
+        factory = TestFactory()
+
+        with requests_mock.Mocker() as m:
+            m.register_uri(requests_mock.ANY, requests_mock.ANY)
+
+            with APIClient.configure_from(factory) as client:
+                client.request(method, "https://from-factory.example.com/foo")
+
+        self.assertEqual(len(m.request_history), 1)
+        self.assertEqual(m.last_request.url, "https://from-factory.example.com/foo")
+        self.assertEqual(m.last_request.method, method)
+        headers = m.last_request.headers
+        self.assertIn("Authorization", headers)
+        self.assertTrue(headers["Authorization"].startswith("Basic "))
+        self.assertFalse(m.last_request.verify)
+        self.assertEqual(m.last_request.timeout, 20.0)
+
+    @given(http_methods)
+    def test_relative_urls_are_made_absolute(self, method):
+        factory = TestFactory()
+        client = APIClient.configure_from(factory)
+
+        with requests_mock.Mocker() as m, client:
+            m.register_uri(requests_mock.ANY, requests_mock.ANY)
+
+            client.request(method, "foo")
+
+        self.assertEqual(len(m.request_history), 1)
+        self.assertEqual(m.last_request.url, "https://from-factory.example.com/foo")
+
+    @given(http_methods)
+    def test_absolute_urls_must_match_base_url(self, method):
+        factory = TestFactory()
+        client = APIClient.configure_from(factory)
+
+        with self.assertRaises(InvalidURLError):
+            client.request(method, "https://example.com/bar")
+
+    @given(http_methods)
+    def test_absolute_urls_must_match_base_url_happy_flow(self, method):
+        factory = TestFactory()
+        client = APIClient.configure_from(factory)
+
+        with requests_mock.Mocker() as m, client:
+            m.register_uri(requests_mock.ANY, requests_mock.ANY)
+
+            client.request(method, "https://from-factory.example.com/foo/bar")
+
+        self.assertEqual(len(m.request_history), 1)
+        self.assertEqual(m.last_request.url, "https://from-factory.example.com/foo/bar")
+
+    @given(http_methods)
+    def test_discouraged_usage_without_context(self, method):
+        client = APIClient("https://example.com")
+
+        with requests_mock.Mocker() as m:
+            m.register_uri(requests_mock.ANY, requests_mock.ANY)
+
+            with patch.object(client, "close", wraps=client.close) as mock_close:
+                client.request(method, "foo")
+
+        self.assertEqual(len(m.request_history), 1)
+        mock_close.assert_called_once()
+
+    @given(http_methods)
+    def test_encouraged_usage_with_context_do_not_close_prematurely(self, method):
+        client = APIClient("https://example.com")
+
+        with patch.object(client, "close", wraps=client.close) as mock_close:
+            with requests_mock.Mocker() as m, client:
+                m.register_uri(requests_mock.ANY, requests_mock.ANY)
+
+                client.request(method, "foo")
```

Comment:
Either go full nesting (for backwards compatibility) or use the modern way; on first pass my eye glanced over `client` being used as a context in this mix.
```suggestion
        with (
            patch.object(client, "close", wraps=client.close) as mock_close,
            requests_mock.Mocker() as m,
            client,
        ):
                m.register_uri(requests_mock.ANY, requests_mock.ANY)

                client.request(method, "foo")
```